packages/wdio-utils/src/shim.ts

```diff
@@ -99,6 +146,121 @@ let wrapCommand = function wrapCommand<T>(commandName: string, fn: Function): (.
+                            target,
+                            function (this: never, mapIterator: Function): any {
+                                // @ts-ignore
+                                return iterators[prop](this, mapIterator)
+                            },
+                            [mapIterator]
+                        )
+                    }
+
+                    /**
+                     * allow to grab the length of fetch element set, e.g.:
+                     * ```js
+                     * const elemAmount = await $$('foo').length
+                     * ```
+                     */
+                    if (ELEMENT_PROPS.includes(prop)) {
+                        return target.then((res) => res[prop])
+                    }
+
+                    /**
+                     * allow to resolve an chained element query, e.g.:
+                     * ```js
+                     * const elem = await $('foo').$('bar')
+                     * console.log(elem.selector) // "bar"
+                     * ```
+                     */
+                    if (prop === 'then') {
+                        return target[prop].bind(target)
+                    }
+
+                    /**
+                     * call a command on an element query, e.g.:
+                     * ```js
+                     * const tagName = await $('foo').$('bar').getTagName()
+                     * ```
+                     */
+                    return (...args: any[]) => target.then(async (elem) => {
+                        if (!elem) {
+                            let errMsg = 'Element could not be found'
+                            const prevElem = await promise
+                            if (Array.isArray(prevElem) && prevInnerArgs && prevInnerArgs.prop === 'get') {
+                                errMsg = `Index out of bounds! $$(${prevInnerArgs.args[0]}) returned only ${prevElem.length} elements.`
+                            }
+
+                            throw new Error(errMsg)
+                        }
+                        return elem[prop](...args)
+                    })
+                }
+            }
+        )
+    }
+
+    return function (this: Clients.Browser, ...args: any[]) {
+        /**
+         * use sync mode if:
+         * - @wdio/sync package is installed and can be resolved
+         * - we are in a fiber context (flag is set when outer function is wrapped into fibers context)
+         *
+         * also if we run command asynchronous and the command suppose to return an element, we
+         * apply `chainElementQuery` to allow chaining of these promises.
+         */
+        const command = hasWdioSyncSupport && wdioSync && !runAsync
+            ? wdioSync!.wrapCommand(commandName, fn)
+            : ELEMENT_QUERY_COMMANDS.includes(commandName)
+                ? function chainElementQuery(this: Promise<Clients.Browser>, ...args: any[]): any {
```

Comment:
I suggest to either define the function outside of this or to create an arrow function instead